tests/cancun/eip6780_selfdestruct/test_selfdestruct.py

```diff
@@ -0,0 +1,176 @@
+"""
+abstract: Tests [EIP-6780: SELFDESTRUCT only in same transaction](https://eips.ethereum.org/EIPS/eip-6780)
+
+    Tests for [EIP-6780: SELFDESTRUCT only in same transaction](https://eips.ethereum.org/EIPS/eip-6780).
+
+"""  # noqa: E501
+
+from typing import Any, Dict, List
+
+import pytest
+
+from ethereum_test_tools import Account, Environment, StateTestFiller, TestAddress, Transaction
+from ethereum_test_tools.vm.opcode import Opcodes as Op
+
+REFERENCE_SPEC_GIT_PATH = "EIPS/eip-6780.md"
+REFERENCE_SPEC_VERSION = "2f8299df31bb8173618901a03a8366a3183479b0"
+
+SELFDESTRUCT_EIP_NUMBER = 6780
+
+
+@pytest.mark.parametrize(
+    "eips",
+    [
+        [SELFDESTRUCT_EIP_NUMBER],
+        [],
+    ],
+    ids=["eip-enabled", "eip-disabled"],
+)
+@pytest.mark.valid_from("Shanghai")
+def test_create_selfdestruct_same_tx(state_test: StateTestFiller):
+    """
+    TODO test that if a contract is created and then selfdestructs in the same
+    transaction the contract should not be created.
+    """
+    env = Environment(
+        coinbase="0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
+        difficulty=0x20000,
+        gas_limit=10000000000,
+        number=1,
+        timestamp=1000,
+    )
+    post: Dict[Any, Any] = {}
+
+    pre = {
+        TestAddress: Account(balance=1000000000000000000000),
+    }
+
+    """
+    original code
+    test_ops = [
+        "6032",
+        "601c",
+        "6000",
+        "39",  # codecopy
+        # create account
+        "6032",
+        "6000",
+        "6000",  # zero value
+        "f0"
+        # TODO call created account
+        "6000",
+        "6000",
+        "6000",
+        "6000",
+        "6000",
+        "85",  # dup the returned address
+        "45",  # pass gas limit
+        "f1",
+        "00"  # STOP
+        # payload:
+        # copy inner payload to memory
+        "6008",
+        "600c",
+        "6000",
+        "39",  # codecopy
+        # return payload
+        "6008",
+        "6000",
+        "f3",
+        # inner payload:
+        "60016000556000ff",
+    ]
+    """
+    code = Op.CODECOPY(0, 0x1C, 0x32)
+    # create account
+    code += Op.CREATE(0, 0, 0x32)
+    # TODO call created account
+    code += Op.PUSH0
+    code += Op.PUSH0
+    code += Op.PUSH0
+    code += Op.PUSH0
+    code += Op.PUSH0
+    code += Op.DUP6  # dup the returned address
+    code += Op.GASLIMIT
+    code += Op.CALL()
+    code += Op.STOP
+    # payload:
+    # copy inner payload to memory
+    code += Op.CODECOPY(0, 0x0C, 0x08)
+    # return payload
+    # inner payload:
+    code += Op.RETURN(0x00, 0x08)
+    code += Op.SSTORE(0, 1)
+    code += Op.PUSH0
+    code += Op.SELFDESTRUCT
+
+    post["0x5fef11c6545be552c986e9eaac3144ecf2258fd3"] = Account.NONEXISTENT
```

Comment:
The offsets for the codecopy operations are off, an intrinsic risk in bytecode authoring.  I think a Yul version would be easier to read and more resiliant to changes.

```suggestion
def test_create_selfdestruct_same_tx(state_test: StateTestFiller, yul: YulCompiler):
    """
    TODO test that if a contract is created and then selfdestructs in the same
    transaction the contract should not be created.
    """
    env = Environment(
        coinbase="0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
        difficulty=0x20000,
        gas_limit=10000000000,
        number=1,
        timestamp=1000,
    )
    post: Dict[Any, Any] = {}

    pre = {
        TestAddress: Account(balance=1000000000000000000000),
    }
    code = yul(
        """
        object "Outer" {
            code {
                let size := datasize("Middle")
                datacopy(0, dataoffset("Middle"), size)
                let contract := create(0, 0, size)
                pop(call(gas(), contract, 0, 0, 0, 0, 0))
                selfdestruct(0)
            }
            object "Middle" {
                code {
                    let size := datasize("Inner")
                    datacopy(0, dataoffset("Inner"), size)
                    return(0, size)
                }
                object "Inner" {
                    code { selfdestruct(0) }
                }
            }
        }
        """
    )

    post["0x6295ee1b4f6dd65047762f924ecd367c17eabf8f"] = Account.NONEXISTENT
    post["0x64e2ebd6405af8cb348aec519084d3fff42ebba6"] = Account.NONEXISTENT
```

Reply:
Hey @shemnon, thanks for the suggestion :)  I was finishing up the refactor when I saw your comment.

This is how it ended up:
https://github.com/jwasinger/execution-spec-tests/blob/8d1f4ff795a6cc2d47f89f0e29e29fd6495c260a/tests/cancun/eip6780_selfdestruct/test_selfdestruct.py#L74-L90
and
https://github.com/jwasinger/execution-spec-tests/blob/8d1f4ff795a6cc2d47f89f0e29e29fd6495c260a/tests/cancun/eip6780_selfdestruct/test_selfdestruct.py#L140-L153

The first code extract is using an `Initcode` construct that's written in python and wraps up any bytecode in the appropriate initcode, taking into consideration the bytecode length and other things.

The second one, I ended up using extcodecopy to make it easier to copy the initcode to the self-destructing contract.

I've also added several other test cases, but it's not by any means exhaustive yet, so any feedback is greatly appreciated!